Over 2X fromBytes performance increase

folkertdev opened the pull request:
Using some stuff I learned over recent months to optimize even further. 

The basic ideas are 
- make as much progress in one step of `Decode.loop` as possible. In this case by decoding 18 bytes at once
- pay very close attention to allocation and GC
- limit the number of function calls

folkertdev committed:
roughly 2X speed improvement in fromBytes

By bundling decoding, doing 18 bytes in one iteration, decoding can be
even faster. Saves jumps and allocation of `Loop _`

folkertdev committed:
Optimize all the way

Look more closely at allocation and number of function calls

By extracting functions to the top-level, they are not re-created upon
every call. That matters for GC

By manually inlining/combining functions, the total number of function
calls goes down. The decoding is suffiently intensive that number of
function calls becomes a problem. Over 30% of time is spent in A2, A4,
etc.

danfishgold commented on src/Decode.elm:
```diff
@@ -151,3 +160,124 @@ unsafeToChar n =
+        w =
+            unsafeToChar (Bitwise.and bits3 lowest6BitsMask)
+    in
+    accum
+        |> String.cons s
+        |> String.cons r
+        |> String.cons q
+        |> String.cons p
+        |> String.cons d
+        |> String.cons c
+        |> String.cons b
+        |> String.cons a
+        |> String.cons w
+        |> String.cons z
+        |> String.cons y
+        |> String.cons x
```
Would it be more performant to use `String.fromList` here? If so, please change the documentation in the beginning of `bitsToCharsSpec` to match the change.
Alternatively: what about having `{ remaining: Int, chars: List Char }` instead of `{ remaining: Int, string: String }` as the step type, and then combining the chars in the `Done` branch? Or even better (or worse?) use `List Int` to decode a list of 6 bit integers and then do `int6s |> List.map unsafeToChar |> String.fromList` outside of the decoder?

folkertdev replied:
Most likely a list is significantly worse. Creating a list still has quite a memory cost. For instance `1 + 2 + 3` is much, much faster than `List.sum [1,2,3]`.

But, a lot of function calls happen here, so I'll try and get some hard numbers. It might be that the cost of function calls outweighs list creation/folding here. But I suspect not.

folkertdev replied:
right. I'd put this in an ellie but it still runs 0.19.0 (elm-benchmark triggers the Map.! bug) and this should be ran with `--optimize` anyway. 

Results: using `String.cons` is roughly 2.3 times faster here 

```
old: 1,156,170
new: 3,866,828
change: +234.45%
```

Normally this wouldn't matter of course, but `fromBytes` is constrained by two factors: how fast can the input be read and the output be written. So in this case it does really matter than string creation is fast. I'll link to this comment in the code.

```elm
module Main exposing (main)

import Benchmark exposing (..)
import Benchmark.Runner exposing (BenchmarkProgram, program)


main : BenchmarkProgram
main =
    program suite


suite : Benchmark
suite =
    comp


comp =
    let
        s =
            'a'

        r =
            'a'

        q =
            'a'

        p =
            'a'

        d =
            'a'

        c =
            'a'

        b =
            'a'

        a =
            'a'

        w =
            'a'

        z =
            'a'

        y =
            'a'

        x =
            'a'
    in
    Benchmark.compare "string from chars"
        "old"
        (\_ ->
            String.fromList [ s, r, q, p, d, c, b, a, w, z, y, x ]
        )
        "new"
        (\_ ->
            ""
                |> String.cons s
                |> String.cons r
                |> String.cons q
                |> String.cons p
                |> String.cons d
                |> String.cons c
                |> String.cons b
                |> String.cons a
                |> String.cons w
                |> String.cons z
                |> String.cons y
                |> String.cons x
        )
```

danfishgold commented on src/Decode.elm:
```diff
@@ -151,3 +160,124 @@ unsafeToChar n =
+        |> bitsToCharsSpec combined3 combined2 combined1
+
+
+bitsToCharsSpec : Int -> Int -> Int -> String -> String
```
I'm not sure I understand why you used this name. Would you mind using something more descriptive here?

danfishgold commented on src/Decode.elm:
```diff
@@ -151,3 +160,124 @@ unsafeToChar n =
+        s =
+            unsafeToChar (Bitwise.and bits1 lowest6BitsMask)
+    in
+    let
```
Are these nested `let...in` blocks more efficient? If so, please add some documentation to explain it. Also: that's really cool and interesting!

folkertdev replied:
they are a visual separation. They compile away completely. There are 3 different parts here and I like the distinction. A comment might work too, if you prefer that?

danfishgold commented on src/Decode.elm:
```diff
@@ -106,26 +112,29 @@ bitsToChars bits missing =
     case missing of
+        0 ->
+            String.cons p (String.cons q (String.cons r (String.fromChar s)))
+
```
Why this change?

folkertdev replied:
the assumption is that this is the case you hit almost all the time. It now requires only one comparison.

danfishgold reviewed (changes requested):
Fantastic work, @folkertdev! I left a couple of comments about missing documentation because I think these new changes make the code harder to parse (even more so than before 😅) and I'd like to keep it relatively understandable

folkertdev committed:
Improve documentation

Optimized code can look funny, and speed must be balanced with
maintainability. So I tried to explain why certain decisions were made.

folkertdev commented:
You're very right to point out maintainability. I've tried to add some clarifications. If anything else is unclear please let me know.

folkertdev committed:
Further documentation tweaks

danfishgold reviewed (approved):
Thanks again, @folkertdev!
I'll publish a new version later today
